---
title: "BCB420- A2"
author: "Mit Patel"
output:
  html_document:
    df_print: paged
---
```{r nest a1, message=FALSE, warning=FALSE, child='a1_notext.Rmd', include=FALSE, eval=TRUE}

```

# A2 starts here
```{r setup a2, message=FALSE, results='hide'}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
   install.packages(pkgs = c("BiocManager"),
           repos = "http://cran.rstudio.org",
           dependencies = TRUE,
           quiet = TRUE) 
}
if (!requireNamespace("circlize", quietly = TRUE)) {
   install.packages(pkgs = c("circlize"),
           repos = "http://cran.rstudio.org",
           dependencies = TRUE,
           quiet = TRUE) 
}

if (!requireNamespace("gprofiler2", quietly = TRUE)) {
   install.packages(pkgs = c("gprofiler2"),
           repos = "http://cran.rstudio.org",
           dependencies = TRUE,
           quiet = TRUE) 
}

BiocManager::install("ComplexHeatmap")
BiocManager::install("limma")
BiocManager::install("GOstats")

library(Biobase)
library(circlize)
library(limma)
library(grid)
library(gprofiler2)

normalized_count_data <- normalized_counts_annot_filtered
```
```{r}

intron_cols <- grep("Intron.*|ensembl_gene_id|hgnc_symbol*", names(normalized_count_data), value = TRUE)
exon_cols <- grep("Exon*|ensembl_gene_id|hgnc_symbol", names(normalized_count_data), value = TRUE)
normalized_count_data_introns <- normalized_count_data[ , intron_cols]
normalized_count_data_exons <- normalized_count_data[ , exon_cols]

heatmap_matrix <- normalized_count_data[ ,3:ncol(normalized_count_data)]
rownames(heatmap_matrix) <- normalized_count_data$ensemble_gene_id
colnames <- colnames(normalized_count_data[ ,3:ncol(normalized_count_data)])


heatmap_matrix_intron <- normalized_count_data_introns[ ,3:ncol(normalized_count_data_introns)]
rownames(heatmap_matrix_intron) <- normalized_count_data_introns$ensemble_gene_id
colnames <- colnames(normalized_count_data_introns[ ,3:ncol(normalized_count_data_introns)])

heatmap_matrix_exon <- normalized_count_data_exons[ ,3:ncol(normalized_count_data_exons)]
rownames(heatmap_matrix_intron) <- normalized_count_data_exons$ensemble_gene_id
colnames <- colnames(normalized_count_data_exons[ ,3:ncol(normalized_count_data_exons)])
```

Note: Code for Heatmap matrix for unthresholded dataset is commented out and hidden as it takes too long to execture. 
```{r heatmap prefilter, echo=FALSE}
# get_heatmap <- function(df) {
#   heatmap_matrix <- df[ ,3:ncol(df)]
#   rownames(heatmap_matrix) <- df$ensemble_gene_id
#   colnames <- colnames(df[ ,3:ncol(df)])  
#   if(min(heatmap_matrix) == 0){
#       heatmap_col = colorRamp2(c( 0, max(heatmap_matrix)), c( "white", "red"))
#     } else {
#       heatmap_col = colorRamp2(c(min(heatmap_matrix), 0, max(heatmap_matrix)), c("blue", "white", "red"))
#     }
#   current_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix),
#                                  show_row_dend = TRUE,
#                                  show_column_dend = TRUE, 
#                                  col=heatmap_col,
#                                  show_column_names = TRUE, 
#                                  show_row_names = FALSE,
#                                  show_heatmap_legend = TRUE
#                                  )
#   current_heatmap
# }
# get_heatmap(normalized_count_data_introns)
# get_heatmap(normalized_count_data_exons)

```
```{r heatmap normalized prefilter, echo=FALSE}
# get_heatmap_normalized <- function(df) {
#   heatmap_matrix <- df[ ,3:ncol(df)]
#   rownames(heatmap_matrix) <- df$ensemble_gene_id
#   colnames <- colnames(df[ ,3:ncol(df)])  
#   heatmap_matrix <- t(scale(t(heatmap_matrix)))
#   if(min(heatmap_matrix) == 0){
#       heatmap_col = colorRamp2(c( 0, max(heatmap_matrix)), c( "white", "red"))
#     } else {
#       heatmap_col = colorRamp2(c(min(heatmap_matrix), 0, max(heatmap_matrix)), c("blue", "white", "red"))
#     }
#   current_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix),
#                                  show_row_dend = TRUE,
#                                  show_column_dend = TRUE, 
#                                  col=heatmap_col,
#                                  show_column_names = TRUE, 
#                                  show_row_names = FALSE,
#                                  show_heatmap_legend = TRUE
#                                  )
#   current_heatmap
# }
# get_heatmap_normalized(normalized_count_data_introns)
# get_heatmap_normalized(normalized_count_data_exons)
```

```{r}
gene_of_interest <- which(normalized_count_data_exons$hgnc_symbol == "CLOCK")
participants <- c("Exon_15", "Exon_3", "Exon_2","Exon_6", "Exon_7", "Exon_9", "Exon_13", "Exon_4", "Exon_12", "Exon_1")
timepoint_clock_matrix <- matrix(nrow=10, ncol=0)
row.names(timepoint_clock_matrix) <- participants
create_timepoint_matrix <- function(grep_pattern, timepoint) {
  tp <- grep(colnames(normalized_count_data_exons),
                          pattern=grep_pattern)
  m <- (t(normalized_count_data_exons
                       [gene_of_interest, tp]))
  temp_rownames <- strsplit(rownames(m), " ")
  new_rownames <- lapply(temp_rownames, function(x) {
    new_name <- unlist(strsplit(x, "_"))[1:2]
    new_name <- paste(new_name[1], new_name[2], sep="_")
    return(new_name)
  })
  colnames(m) <- c(timepoint)
  row.names(m) <- new_rownames
  return(m)
}
t1 <- create_timepoint_matrix("Exon.*_12.00", "12.00")
t2 <- create_timepoint_matrix("Exon.*_16.00", "16.00")
t3 <- create_timepoint_matrix("Exon.*_20.00", "20.00")
t4 <- create_timepoint_matrix("Exon.*_00.00", "00.00")
t5 <- create_timepoint_matrix("Exon.*_04.00", "04.00")
t6 <- create_timepoint_matrix("Exon.*_08.00", "08.00")
matricies <- list(t1, t2, t3, t4, t5, t6)
for (l in matricies) {
  timepoint_clock_matrix <- cbind(timepoint_clock_matrix, l[, 1][match(rownames(timepoint_clock_matrix), rownames(l))])
}
colnames(timepoint_clock_matrix) <- c("12.00", "16.00", "20.00", "00.00", "04.00", "08.00")
timepoint_clock_matrix
```
 
```{r}
df_clock_timepoint <- data.frame()
for (col in 1:ncol(timepoint_clock_matrix)) {
  values <- as.data.frame(timepoint_clock_matrix[col, ])
  timepoints <- rep(c(colnames(timepoint_clock_matrix)[col]), 6)
  values <- cbind(values, timepoints)
  colnames(values) <- c("values", "timepoints")
  rownames(values) <- NULL
  df_clock_timepoint <- rbind(df_clock_timepoint, values)
}
df_clock_timepoint[1:5, ]
aov_clock <- aov(df_clock_timepoint$values ~ df_clock_timepoint$timepoints)
summary(aov_clock)
```

```{r}
pat_colors <- rainbow(10)
pat_colors <- unlist(lapply(pat_colors,FUN=function(x){rep(x,2)}))
limma::plotMDS(heatmap_matrix_exon,
               col = pat_colors)
```

```{r}
samples <- data.frame(
        lapply(colnames(normalized_count_data_exons)[3:ncol(normalized_count_data_exons)], 
        FUN=function(x){
          unlist(strsplit(x, split = "_"))[c(2,4)]}))
colnames(samples) <- colnames(normalized_count_data_exons)[3:ncol(normalized_count_data_exons)]
rownames(samples) <- c("patient","timepoint")
samples <- data.frame(t(samples))
samples[1:5, ]

```

```{r}
model_design <- model.matrix(~ samples$timepoint)
model_design[1:5, 1:6]
```
```{r}
expressionMatrix <- as.matrix(normalized_count_data_exons[,3:ncol(normalized_count_data_exons)])
rownames(expressionMatrix) <- normalized_count_data_exons$ensembl_gene_id
colnames(expressionMatrix) <- colnames(normalized_count_data_exons)[3:ncol(normalized_count_data_exons)]
minimalSet <- ExpressionSet(assayData=expressionMatrix)
fit <- lmFit(minimalSet, model_design)
fit2 <- eBayes(fit, trend=TRUE)
topfit <- topTable(fit2, 
                   coef=ncol(model_design),
                   adjust.method = "BH",
                   number = nrow(expressionMatrix))

#merge hgnc symboles to the topfit table
output_hits <- merge(normalized_count_data_exons[,1:2],
                     topfit,
                     by.y=0,by.x=1,
                     all.y=TRUE)
#sort the output lists by decreasing pvaluee
output_hits <- output_hits[order(output_hits$P.Value),]

output_hits[1:5, ]

genes_pass <- length(which(output_hits$P.Value < 0.05))
genes_pass_correction <- length(which(output_hits$adj.P.Val < 0.05))
```


```{r}
xlim <- c(-3,1)
ylim <- c(-4,4)
passed_genes <- topfit[which(topfit$P.Value<0.05), ]
not_passed_genes <- topfit[which(topfit$P.Value>=0.05), ]
limma::plotMA(topfit, main="MA Plot: Passed Threshold Genes", ylim=ylim, xlim=xlim)
limma::plotMA(not_passed_genes, main="MA Plot: Other Genes", ylim=ylim, xlim=xlim)

```

```{r}
heatmap_matrix_new <- normalized_count_data_exons[,3:ncol(normalized_count_data_exons)]
rownames(heatmap_matrix_new) <- normalized_count_data_exons$ensembl_gene_id
colnames(heatmap_matrix_new) <- colnames(normalized_count_data_exons[,3:ncol(normalized_count_data_exons)])
heatmap_matrix_new <- t(scale(t(heatmap_matrix_new)))

top_hits <- output_hits$ensembl_gene_id[output_hits$P.Value<0.05]
heatmap_matrix_tophits <- t(
  scale(t(heatmap_matrix_new[
    which(rownames(heatmap_matrix_new) %in% top_hits),])))


if(min(heatmap_matrix_tophits) == 0){
    heatmap_col = circlize::colorRamp2(c( 0, max(heatmap_matrix_tophits)), 
                             c( "white", "red"))
  } else {
    heatmap_col = circlize::colorRamp2(c(min(heatmap_matrix_tophits), 0, max(heatmap_matrix_tophits)), c("blue", "white", "red"))
  }
current_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix_tophits),
                           cluster_rows = TRUE,
                           cluster_columns = TRUE,
                               show_row_dend = TRUE,
                               show_column_dend = FALSE, 
                               show_column_names = TRUE, 
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE,
                           column_title =  "Heatmap of top gene hits - Limma Model",
                                                       column_names_gp = gpar(fontsize = 8)

                               )
current_heatmap
```

```{r}
exon_data_matrix <- inverse_data_matrix[ , grepl("Exon.*", colnames(inverse_data_matrix))]
d = DGEList(counts=exon_data_matrix, group=samples$timepoint)
d <- estimateDisp(d, model_design)
fit <- glmQLFit(d, model_design)
model_design[1:10, 1:5]
qlf.pos_vs_neg <- glmQLFTest(fit)
topTags(qlf.pos_vs_neg)
```

```{r}
qlf_output_hits <- topTags(qlf.pos_vs_neg,sort.by = "PValue",
                           n = nrow(normalized_count_data))
quasi_pvalue_pass <- length(which(qlf_output_hits$table$PValue < 0.05))
quasi_corrected_pass <- length(which(qlf_output_hits$table$FDR < 0.05))

```

```{r}
qlf_pat_model_pvalues <- data.frame(
          ensembl_id = rownames(qlf_output_hits$table),
          qlf_patient_pvalue=qlf_output_hits$table$PValue)
limma_pat_model_pvalues <-  data.frame(
          ensembl_id = output_hits$ensembl_gene_id,
          limma_patient_pvalue = output_hits$P.Value)
two_models_pvalues <- merge(qlf_pat_model_pvalues,
                            limma_pat_model_pvalues,
                            by.x=1,by.y=1)
two_models_pvalues$colour <- "black"
two_models_pvalues$colour[two_models_pvalues$qlf_patient_pvalue<0.05] <- "orange"
two_models_pvalues$colour[two_models_pvalues$limma_patient_pvalue<0.05] <- "blue"
two_models_pvalues$colour[two_models_pvalues$qlf_patient_pvalue<0.05 & two_models_pvalues$limma_patient_pvalue<0.05] <- "red"
plot(two_models_pvalues$qlf_patient_pvalue,
     two_models_pvalues$limma_patient_pvalue,
     col = two_models_pvalues$colour,
     xlab = "QLF patient model p-values",
     ylab ="Limma Patient model p-values",
     main="QLF vs Limma")

```

```{r}
ensembl_of_interest <- normalized_count_data_exons$ensembl_gene_id[
  which(normalized_count_data_exons$hgnc_symbol == "CLOCK")]
two_models_pvalues$colour <- "grey"
two_models_pvalues$colour[two_models_pvalues$ensembl_id==ensembl_of_interest] <- "red"
plot(two_models_pvalues$qlf_patient_pvalue,
     two_models_pvalues$limma_patient_pvalue,
     col = two_models_pvalues$colour,
     xlab = "QLF patient model p-values",
     ylab ="Limma Patient model p-values",
     main="QLF vs Limma")
points(two_models_pvalues[
  two_models_pvalues$ensembl_id==ensembl_of_interest,2:3],
       pch=24,  col="red", cex=1.5)
```

```{r}
top_hits <- rownames(qlf_output_hits$table)[output_hits$P.Value<0.05]
heatmap_matrix_tophits <- t(
  scale(t(heatmap_matrix_new[which(rownames(heatmap_matrix_new) %in% top_hits),])))
if(min(heatmap_matrix_tophits) == 0){
    heatmap_col = circlize::colorRamp2(c( 0, max(heatmap_matrix_tophits)), 
                             c( "white", "red"))
  } else {
    heatmap_col = circlize::colorRamp2(c(min(heatmap_matrix_tophits), 0, max(heatmap_matrix_tophits)), c("blue", "white", "red"))
  }
current_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix_tophits),
                           cluster_rows = TRUE,
                           cluster_columns = TRUE,
                               show_row_dend = TRUE,
                               show_column_dend = TRUE, 
                               col=heatmap_col,
                               show_column_names = TRUE, 
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE,
                               column_title =  "Heatmap of top gene hits - Quasi liklihood Model", 
                            column_names_gp = gpar(fontsize = 8)
)
current_heatmap
```
```{r}
qlf_output_hits_withgn <- merge(normalized_count_data[,1:2],qlf_output_hits, by.x=1, by.y = 0)
qlf_output_hits_withgn[,"rank"] <- -log(qlf_output_hits_withgn$PValue,base =10) * sign(qlf_output_hits_withgn$logFC)
qlf_output_hits_withgn <- qlf_output_hits_withgn[order(qlf_output_hits_withgn$rank),]

upregulated_genes <- qlf_output_hits_withgn$hgnc_symbol[
  which(qlf_output_hits_withgn$PValue < 0.45
             & qlf_output_hits_withgn$logFC > 0)]
downregulated_genes <- qlf_output_hits_withgn$hgnc_symbol[
  which(qlf_output_hits_withgn$PValue < 0.05 
             & qlf_output_hits_withgn$logFC < 0)]

n_upreg_genes <- length(upregulated_genes)
n_downreg_genes <- length(downregulated_genes)
```
We have identified `r {n_upreg_genes}` upregulated genes and `r {n_downreg_genes}` downregulated genes to be used in our gene enrichment analysis. 


```{r}
gostres_upreg <- gprofiler2::gost(upregulated_genes, organism = "hsapiens", ordered_query = FALSE,
  multi_query = FALSE, significant = TRUE, exclude_iea = TRUE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = c("g_SCS", "bonferroni",
  "fdr", "false_discovery_rate", "gSCS", "analytical"),
  domain_scope = c("annotated"), custom_bg = NULL,
  numeric_ns = "", sources = NULL)

gostres_downreg <- gprofiler2::gost(downregulated_genes, organism = "hsapiens", ordered_query = FALSE,
  multi_query = FALSE, significant = TRUE, exclude_iea = TRUE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = 0.05, correction_method = c("g_SCS", "bonferroni",
  "fdr", "false_discovery_rate", "gSCS", "analytical"),
  domain_scope = c("annotated"), custom_bg = NULL,
  numeric_ns = "", sources = NULL)
```

```{r}
gprofiler2::gostplot(gostres_upreg, capped = TRUE, interactive = TRUE, pal = c(`GO:MF`
  = "#dc3912", `GO:BP` = "#ff9900", `GO:CC` = "#109618", KEGG = "#dd4477",
  REAC = "#3366cc", WP = "#0099c6", TF = "#5574a6", MIRNA = "#22aa99", HPA
  = "#6633cc", CORUM = "#66aa00", HP = "#990099"))

```

```{r}
gprofiler2::gostplot(gostres_downreg, capped = TRUE, interactive = TRUE, pal = c(`GO:MF`
  = "#dc3912", `GO:BP` = "#ff9900", `GO:CC` = "#109618", KEGG = "#dd4477",
  REAC = "#3366cc", WP = "#0099c6", TF = "#5574a6", MIRNA = "#22aa99", HPA
  = "#6633cc", CORUM = "#66aa00", HP = "#990099"))
```